File: Mac/BarkCloudKit/Sources/BarkCloudKit/Data/Cloud/DynamicFolderRepository.swift
import Foundation
import GRPCCore
import SwiftProtobuf

/// Доступ к сервису умных (динамических) папок (`DynamicFolderApi`). Зеркалит
/// `AlbumRepository`: доменные ошибки пробрасываются как `RPCError`, UI маппит их
/// через `domainErrorMessage(_:)`. Системные папки приходят в списке первыми.
public final class DynamicFolderRepository: Sendable {
    private let grpc: GrpcManager

    public init(grpc: GrpcManager) { self.grpc = grpc }

    /// Системные + пользовательские папки (с обложкой/счётчиком).
    public func listFolders() async throws -> [DynamicFolderCard] {
        let stub = try await grpc.dynamicFolderStub()
        let resp = try await stub.listDynamicFolders(Barkcloud_Files_ListDynamicFoldersRequest())
        return resp.folders.map(DynamicFolderCard.init)
    }

    /// Содержимое папки по критериям (cursor-пагинация, опциональный фильтр по типу).
    public func listItems(
        folderID: String,
        kindFilter: CloudMediaKind? = nil,
        limit: Int32 = 50,
        cursorCreatedAt: Date? = nil,
        cursorFileID: String = ""
    ) async throws -> DynamicFolderItemsPage {
        let stub = try await grpc.dynamicFolderStub()
        var req = Barkcloud_Files_ListDynamicFolderItemsRequest()
        req.folderID = folderID
        req.limit = limit
        if let cursorCreatedAt {
            req.cursorCreatedAt = Google_Protobuf_Timestamp(date: cursorCreatedAt)
        }
        req.cursorFileID = cursorFileID
        if kindFilter == .video {
            req.kindFilter = .video
        } else if kindFilter == .photo {
            req.kindFilter = .photo
        }
        let resp = try await stub.listDynamicFolderItems(req)
        return DynamicFolderItemsPage(
            items: resp.items.map { MediaAsset($0.file) },
            nextCursorCreatedAt: resp.hasNextCursorCreatedAt ? resp.nextCursorCreatedAt.date : nil,
            nextCursorFileID: resp.nextCursorFileID
        )
    }

    @discardableResult
    public func create(
        name: String,
        combinator: Barkcloud_Files_DfCombinator,
        rules: [DynamicFolderRule],
        viewMode: Barkcloud_Files_DfViewMode
    ) async throws -> DynamicFolderCard {
        let stub = try await grpc.dynamicFolderStub()
        var req = Barkcloud_Files_CreateDynamicFolderRequest()
        req.name = name
        req.combinator = combinator
        req.rules = rules.map(\.proto)
        req.viewMode = viewMode
        return DynamicFolderCard(try await stub.createDynamicFolder(req))
    }

    /// Обновление: `name`/`viewMode` — optional-поля (заданы = «менять»); правила и
    /// комбинатор всегда заменяются целиком (как в вебе).
    @discardableResult
    public func update(
        folderID: String,
        name: String,
        combinator: Barkcloud_Files_DfCombinator,
        rules: [DynamicFolderRule],
        viewMode: Barkcloud_Files_DfViewMode
    ) async throws -> DynamicFolderCard {
        let stub = try await grpc.dynamicFolderStub()
        var req = Barkcloud_Files_UpdateDynamicFolderRequest()
        req.folderID = folderID
        req.name = name
        req.combinator = combinator
        req.rules = rules.map(\.proto)
        req.viewMode = viewMode
        return DynamicFolderCard(try await stub.updateDynamicFolder(req))
    }

    public func delete(folderID: String) async throws {
        let stub = try await grpc.dynamicFolderStub()
        var req = Barkcloud_Files_DeleteDynamicFolderRequest()
        req.folderID = folderID
        _ = try await stub.deleteDynamicFolder(req)
    }
}

File: Android/BarkCloud.Android/app/src/main/java/com/barkfluff/BarkCloud/data/cloud/DynamicFolderRepository.kt
package com.barkfluff.BarkCloud.data.cloud

import barkcloud.files.FilesApiOuterClass.CreateDynamicFolderRequest
import barkcloud.files.FilesApiOuterClass.DeleteDynamicFolderRequest
import barkcloud.files.FilesApiOuterClass.DfCombinator
import barkcloud.files.FilesApiOuterClass.DfViewMode
import barkcloud.files.FilesApiOuterClass.ListDynamicFolderItemsRequest
import barkcloud.files.FilesApiOuterClass.ListDynamicFoldersRequest
import barkcloud.files.FilesApiOuterClass.MediaKind
import barkcloud.files.FilesApiOuterClass.UpdateDynamicFolderRequest
import com.barkfluff.BarkCloud.grpc.GrpcManager

class DynamicFolderRepository(
    private val grpc: GrpcManager,
) {

    suspend fun listFolders(): List<DynamicFolderCard> {
        val resp = grpc.dynamicFolderStub()
            .listDynamicFolders(ListDynamicFoldersRequest.getDefaultInstance())
        return resp.foldersList.map(DynamicFolderCard::from)
    }

    suspend fun listItems(
        folderId: String,
        limit: Int = 60,
        cursorCreatedAtMillis: Long? = null,
        cursorFileId: String = "",
    ): DynamicFolderItemsPage {
        val req = ListDynamicFolderItemsRequest.newBuilder()
            .setFolderId(folderId)
            .setLimit(limit)
            .setCursorFileId(cursorFileId)
        if (cursorCreatedAtMillis != null) req.setCursorCreatedAt(cursorCreatedAtMillis.toTimestamp())
        val resp = grpc.dynamicFolderStub().listDynamicFolderItems(req.build())
        return DynamicFolderItemsPage(
            items = resp.itemsList.map { MediaAsset.from(it.file) },
            nextCursorCreatedAtMillis = if (resp.hasNextCursorCreatedAt()) resp.nextCursorCreatedAt.toEpochMillis() else null,
            nextCursorFileId = resp.nextCursorFileId,
        )
    }

    suspend fun create(
        name: String,
        combinator: DfCombinator,
        rules: List<DynamicFolderRule>,
        viewMode: DfViewMode,
    ): DynamicFolderCard {
        val resp = grpc.dynamicFolderStub().createDynamicFolder(
            CreateDynamicFolderRequest.newBuilder()
                .setName(name)
                .setCombinator(combinator)
                .addAllRules(rules.map { it.toProto() })
                .setViewMode(viewMode)
                .build()
        )
        return DynamicFolderCard.from(resp)
    }

    suspend fun update(
        folderId: String,
        name: String,
        combinator: DfCombinator,
        rules: List<DynamicFolderRule>,
        viewMode: DfViewMode,
    ): DynamicFolderCard {
        val resp = grpc.dynamicFolderStub().updateDynamicFolder(
            UpdateDynamicFolderRequest.newBuilder()
                .setFolderId(folderId)
                .setName(name)
                .setCombinator(combinator)
                .addAllRules(rules.map { it.toProto() })
                .setViewMode(viewMode)
                .build()
        )
        return DynamicFolderCard.from(resp)
    }

    suspend fun delete(folderId: String) {
        grpc.dynamicFolderStub().deleteDynamicFolder(
            DeleteDynamicFolderRequest.newBuilder().setFolderId(folderId).build()
        )
    }
}
